### PR TITLE
Fold sharded support 

### PR DESCRIPTION
**Pull request by yugaoTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/10288

### Problem description
Add sharding support to fold op composited with pad, reshape, transpose.

### What's changed
fix bug in tensor.reshape, by updating the shard spec and page size
add python/c++ versions of fold sharded.

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/10508799876

### CODE REVIEW

**Review by ayerofieiev-tt — COMMENTED** [review]
@xanderchin should we be mindful of sweep infra and extend the sweep record to indicate our expectation that this op supports sharding?

File: tt_metal/impl/buffers/buffer.hpp [hunk 1]
```diff
@@ -184,6 +185,8 @@ class Buffer {
 
     uint32_t num_pages() const { return this->size() / this->page_size(); }
 
+    void set_page_size(uint64_t page_size) { page_size_ = page_size; }
```
Comment (abhullar-tt) [inline tt_metal/impl/buffers/buffer.hpp:188]: should add checks here that the new page size doesnt cause the buffer to exceed previously allocated size

**Review by eyonland — APPROVED** [review]

File: tt_metal/impl/buffers/buffer.hpp [hunk 2]
```diff
@@ -239,6 +242,10 @@ class Buffer {
         return this->shard_parameters_.value();
     }
 
+    void set_shard_spec(ShardSpecBuffer shard_spec) {
```
Comment (abhullar-tt) [inline tt_metal/impl/buffers/buffer.hpp:245]: should this be passed in as const ref?

**Review by abhullar-tt — CHANGES_REQUESTED** [review]

File: tt_metal/impl/buffers/buffer.hpp (hunk 1 above)
Comment (yugaoTT) [reply tt_metal/impl/buffers/buffer.hpp]: yes, changed

File: tt_metal/impl/buffers/buffer.hpp (hunk 2 above)
Comment (yugaoTT) [reply tt_metal/impl/buffers/buffer.hpp]: yes, changed

**Review by abhullar-tt — APPROVED** [review]

**Review by ntarafdar — APPROVED** [review]

**Pull request merged**